### PR TITLE
ARROW-4230: [C++] Fix Flight builds with gRPC/Protobuf/c-ares

### PR DESCRIPTION
**Pull request by ghost**
- Non-vendored, non-CMake gRPC dependencies are accepted via `GRPC_HOME`
- Non-vendored, non-CMake c-ares dependencies are accepted via `CARES_HOME`
- Refactored gRPC dependency configuration into `FindgRPC.cmake`

```shell
# Build with all system dependencies
env GRPC_HOME=/usr/local/Cellar/grpc/1.17.2 CARES_HOME=/usr/local/Cellar/c-ares/1.15.0 PROTOBUF_HOME=/usr/local/Cellar/protobuf/3.6.1.3_1 cmake -DARROW_FLIGHT=ON 
# Build with all vendored dependencies
cmake -DARROW_FLIGHT=ON
# Build with CMake gRPC installation
cmake -DgRPC_INSTALL=ON -DgRPC_BUILD_TESTS=OFF -DgRPC_PROTOBUF_PROVIDER=package -DgRPC_ZLIB_PROVIDER=package -DgRPC_CARES_PROVIDER=package -DgRPC_SSL_PROVIDER=package -DOPENSSL_ROOT_DIR=/usr/local/Cellar/openssl/1.0.2q && make -j
cd path/to/arrow
env GRPC_HOME=/usr/local/lib/cmake/grpc CARES_HOME=/usr/local/ cmake -DARROW_FLIGHT=ON -DgRPC_DIR=/usr/local/lib/cmake/grpc 
```

### CODE REVIEW

**Comment by wesm** [comment]
@lihalite thank you for this, will review. Can you please create a new JIRA? This does not resolve ARROW-3149, which is still not in conda-forge

**Comment by lihalite** [comment]
@wesm of course, created [ARROW-4230](https://issues.apache.org/jira/browse/ARROW-4230).

**Review by pitrou — DISMISSED** [review]
We already have the `GRPC_HOME` environment variable, which is tested in  `cmake_modules/ThirdPartyToolchain.cmake`. Similarly, the `PROTOBUF_HOME` environment variable is already tested as well. Therefore, without giving a detailed review, it looks like some of those changes are unnecessary.

Also, since we are currently using environment variables for this, please don't make a CMake variable for `CARES_HOME`.

**Comment by lihalite** [comment]
Hmm, I think I'm misunderstanding - `GRPC_HOME` exists, but it's used with `find_package(CONFIG)`, so gRPC must be installed with CMake. This adds `FindgRPC.cmake` so that a non-CMake installation will also work. The reason why you see `gRPC_DIR` in the example invocation is because that's a CMake-defined variable for the search path in case you do want to use a CMake-built gRPC.

I missed the part where the `_HOME` variables were set from the environment, so I will add a similar statement for `CARES_HOME` and fix the README section I added.

**Comment by pitrou** [comment]
> This adds FindgRPC.cmake so that a non-CMake installation will also work. 

I'm a bit surprised. I set `RE2_HOME=/usr` to use Ubuntu's system install of re2, yet `find_package` seems to work fine. There does not seem to be any re2-specific cmake file installed.

**Comment by pitrou** [comment]
Ah, but we have a `FindRE2.cmake`. Got it. Yet more boilerplate in the build files ;-)

**Comment by lihalite** [comment]
Yup, unfortunately :/ I think the other difference is that `find_package(gRPC CONFIG REQUIRED)` explicitly looks for a CMake-generated install vs `find_package(RE2 REQURIED)` which only looks for -some- re2 install. And for gRPC we need to know a little more about the package (to get the path to the protoc plugin, for instance).

**Comment by lihalite** [comment]
This now sets `CARES_HOME` from the environment; the README has been updated to use environment variables.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (pitrou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: It would be nice to keep those directives in lexographical order.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (pitrou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: I do not see `CARES_STATIC_LIB` used anywhere in our build files. Am I missing something?

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (lihalite) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: Agreed, fixed :)

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (pitrou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: Is there a reason for commenting those things out? Ideally, we should remove them if not necessary anymore.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (lihalite) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: It's used below in `ADD_THIRDPARTY_LIB` - it's not actually used by Arrow/Flight, but by gRPC.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (lihalite) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: An oversight on my part, thanks.

File: cpp/README.md
Comment (pitrou) [inline cpp/README.md]: Something missing here? On Ubuntu 18.04, one might get away with (untested):
```
apt install libgrpc++-dev
```

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (pitrou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: Ah, sorry :-)

File: cpp/README.md
Comment (lihalite) [reply cpp/README.md]: Looks like I'm being scatterbrained today; I meant to spin up a VM to test this before I added it to the readme.

Comment (lihalite) [reply cpp/README.md]: Looks like you need several packages, and Debian/Ubuntu statically links one of the gRPC libraries into the others, which adds another permutation... the README/FindgRPC now cover this.

Comment (lihalite) [reply cpp/README.md]: aaaand 18.04 ships too old a version of gRPC, let me follow up on this...

Comment (wesm) [reply cpp/README.md]: I'd be a bit shocked if we were able to use Linux system packages. gRPC has been too fast-moving

Comment (lihalite) [reply cpp/README.md]: Fair enough. Maybe I can mention these variables, but say it's at the user's own risk?

**Comment by bissont** [comment]
Thanks for this! Just tested and it does build the arrowflight library with a cmake -DARROW_FLIGHT=ON; make, but if you turn -DARROW_BUILD_TESTS=ON, cmake fails:

CMake Error at src/arrow/flight/CMakeLists.txt:89 (add_dependencies):
  Cannot add target-level dependencies to non-existent target "flight-test".

  The add_dependencies works for top-level logical targets created by the
  add_executable, add_library, or add_custom_target commands.  If you want to
  add file-level dependencies see the DEPENDS option of the add_custom_target
  and add_custom_command commands.


-- Configuring incomplete, errors occurred!

**Comment by lihalite** [comment]
Ah, I had a fix on a separate branch, let me put it here as well.

**Comment by lihalite** [comment]
@bissont give it another try!

**Comment by bissont** [comment]
Sorry for the delay - been trying a few different combinations...

Cmake passes, but make fails because it couldn't find my grpc_cpp_plugin binary beccause it tries to generate the flight protobuf header/c files before the grpc dependency is built:
/home/t.bisson/git/a2/cpp/debug/grpc_ep-prefix/src/grpc_ep-build/./grpc_cpp_plugin: program not found or is not executable                             

I wainted for make to continue and build grpc, then ran make again and it built the protobuf header/c files. Seems like a dependency ordering issue.

----------------------------------------------

After that, it looks like it's missing dependencies like a -lssl or -lcrypt based  dependency is missed becaus I'm seeing undefined references when trying to link flight-test-server.
'SSL_CTX_set_session_id_context' 

[ 99%] Linking CXX executable ../../../debug/flight-test-server
/home/t.bisson/git/a2/cpp/debug/grpc_ep-prefix/src/grpc_ep/third_party/boringssl/include/openssl/x509.h:161: error: undefined reference to 'sk_new_null'
/home/t.bisson/git/a2/cpp/debug/grpc_ep-prefix/src/grpc_ep/third_party/boringssl/include/openssl/x509.h:161: error: undefined reference to 'sk_pop_free'
/home/t.bisson/git/a2/cpp/debug/grpc_ep-prefix/src/grpc_ep/third_party/boringssl/include/openssl/x509.h:161: error: undefined reference to 'sk_push'
/home/t.bisson/git/a2/cpp/debug/grpc_ep-prefix/src/grpc_ep/third_party/boringssl/include/openssl/x509v3.h:214: error: undefined reference to 'sk_num'
/home/t.bisson/git/a2/cpp/debug/grpc_ep-prefix/src/grpc_ep/third_party/boringssl/include/openssl/x509v3.h:214: error: undefined reference to 'sk_value'
/home/t.bisson/git/a2/cpp/debug/grpc_ep-prefix/src/grpc_ep/third_party/boringssl/include/openssl/x509v3.h:214: error: undefined reference to 'sk_pop_free'
/home/t.bisson/git/a2/cpp/debug/grpc_ep-prefix/src/grpc_ep/src/core/tsi/ssl_transport_security.cc:157: error: undefined reference to 'SSL_library_init'
/home/t.bisson/git/a2/cpp/debug/grpc_ep-prefix/src/grpc_ep/src/core/tsi/ssl_transport_security.cc:158: error: undefined reference to 'SSL_load_error_strings'

My current build commands are:
cmake -DARROW_FLIGHT=ON -DARROW_BUILD_TESTS  ..  && make -j 64

**Comment by lihalite** [comment]
Hmm, that sounds like a problem in the gRPC build process; it might be worth building gRPC separately then.

**Comment by bissont** [comment]
Yes, using a separate gRPC works. This was my command:
GRPC_HOME=/usr/local/ CARES_HOME=/usr/local/ PROTOBUF_HOME=/usr/local cmake -DARROW_FLIGHT=ON -DARROW_BUILD_TESTS=ON .. && make

**Comment by wesm** [comment]
@lihalite we are a bit jammed up right now with the 0.12 release. I wanted to take the build for a spin but I can also do that after this is merged, this seems like a significant improvement over the current state of things

**Comment by lihalite** [comment]
@wesm ah no worries, thanks for the update - totally understand that everyone's busy with the release.

**Comment by pitrou** [comment]
@lihalite I'll take a look soon.

**Comment by lihalite** [comment]
Of course, just wanted to follow up. Thanks for the update.

**Comment by fsaintjacques** [comment]
I'll also provide input on the build.

**Comment by wesm** [comment]
OK, well out of @pitrou @fsaintjacques and myself, only one of us should actually take time to test this out. Are you volunteering @fsaintjacques? If so I'll defer to you

**Comment by fsaintjacques** [comment]
@wesm depends, are you testing on UB14 or UB18? I'd say the more different linux distros we get, the better.

**Comment by wesm** [comment]
Since we don't run Flight in CI yet I think we should avoid being significant time on this right now. We'll work out further build problems as this moves closer to production-readiness

**Comment by wesm** [comment]
It definitely would be useful (adding gRPC to conda-forge). That's a pre-requisite

**Review by wesm — APPROVED** [review]
+1. I added a couple tweaks for c-ares because the static library it produces is called `libcares_static.a`. I built gRPC 1.18.0 (with some effort) using dependencies from conda-forge and using the `Makefile` in grpc instead of CMake and got things to work. The gRPC Makefile has some weird behavior -- for example, it strips symbols from the static libraries causing linker failures here. I also found that I had to hack out `-Werror` since there were compiler warnings with gcc 7.3.0. Overall a pretty difficult experience -- we should try to limit the amount of pain and suffering we are foisting on users of the Arrow libraries by getting the packages on conda-forge or other package managers soon

**Comment by lihalite** [comment]
Thanks for the review! I've got some integration tests/compatibility fixes in the queue (I've commented on [ARROW-4213](https://issues.apache.org/jira/browse/ARROW-4213)) and should be able to look at the conda-forge stuff, so that we can get this all in CI.

**Comment by wesm** [comment]
sounds good. thanks @lihalite!

**Pull request closed without merging**
